### PR TITLE
[PLA-3620] Feature/molecular structure featurizer

### PR DESCRIPTION
**Pull request by sparadiso**
# Citrine Python PR

## Description 
Add molecular formula featurizer.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [ ] I have added tests for 100% coverage
- [ ] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [ ] I have bumped the version in setup.py

### CODE REVIEW

File: src/citrine/informatics/predictors.py [hunk 1]
```diff
@@ -272,3 +276,122 @@ def __str__(self):
...
+        name of the configuration
+    description: str
+        the description of the predictor
+    descriptor: MolecularStructureDescriptor
+        the descriptor to featurize
+    features: List[str]
+        the list of features to compute. Valid values include:
+        - all
+        - standard
+        - expensive
```
Comment (kroenlein) [inline src/citrine/informatics/predictors.py:297]: I assume the first three are sets of features rather than particular ones; probably worth mentioning or inserting an empty line.

File: src/citrine/informatics/predictors.py [hunk 2]
```diff
@@ -272,3 +276,122 @@ def __str__(self):
...
+        - MassAutocorr
+        - AcidGroupCount
+        - BPol
+        - WienerNumbers
+        - LongestAliphaticChain
+        - CPSA
+        - ChiCluster
+        - BasicGroupCount
+    excludes: List[str]
+        list of features to exclude (accepts same set of values as features)
```
Comment (kroenlein) [inline src/citrine/informatics/predictors.py:347]: Document what happens with values in both lists?

File: tests/informatics/test_predictors.py [hunk 3]
```diff
@@ -24,6 +25,17 @@ def simple_predictor() -> SimpleMLPredictor:
...
 
 
+@pytest.fixture
+def molecule_featurizer() -> MolecularStructureFeaturizer:
+    return MolecularStructureFeaturizer(
+        name="Molecule featurizer",
+        description="description",
+        descriptor=MolecularStructureDescriptor("SMILES"),
+        features=["standard"],
+        excludes=[]
```
Comment (kroenlein) [inline tests/informatics/test_predictors.py:35]: If this is for a test, why aren't we explicitly excluding one of the features to make sure we're exercising all code?

File: tests/informatics/test_predictors.py [hunk 4]
```diff
@@ -110,3 +122,30 @@ def test_expression_post_build(expression_predictor):
...
+    assert molecule_featurizer.name == "Molecule featurizer"
+    assert molecule_featurizer.description == "description"
+    assert molecule_featurizer.descriptor == MolecularStructureDescriptor("SMILES")
+    assert molecule_featurizer.features == ["standard"]
+    assert molecule_featurizer.excludes == []
+
+    assert str(molecule_featurizer) == "<MolecularStructureFeaturizer 'Molecule featurizer'>"
+
+    # This is what you get when you require 100% test coverage
+    assert molecule_featurizer._post_dump(molecule_featurizer.dump())["display_name"] == molecule_featurizer.name
```
Comment (bfolie) [inline tests/informatics/test_predictors.py:137]: Why is this needed for coverage? Shouldn'T `molecule_featurizer.dump()` automatically call the code in `_post_dump()`

Comment (sparadiso) [reply tests/informatics/test_predictors.py:137]: 🤷‍♂️without this, test coverage for _post_dump wasn't detected. I'll check to see if dump() is covered.

File: tests/informatics/test_predictors.py (hunk 3 above)
Comment (sparadiso) [reply tests/informatics/test_predictors.py:35]: Could do - we don't _actually_ test the logic here, though. We don't do set math or anything in the client. I can confirm a non-empty list is preserved, though.

File: src/citrine/informatics/predictors.py (hunk 2 above)
Comment (sparadiso) [reply src/citrine/informatics/predictors.py:347]: You mean explain that excludes means "any feature in this list and also the features list will be excluded"?

File: tests/informatics/test_predictors.py (hunk 4 above)
Comment (sparadiso) [reply tests/informatics/test_predictors.py]: dump() wasn't covered - that did it 👍

File: src/citrine/informatics/predictors.py (hunk 2 above)
Comment (bfolie) [reply src/citrine/informatics/predictors.py:348]: I think that's what he means.

I got asked this question, what happens if a feature is in both lists, when writing up the design doc. So I do think we need to explicitly explain the pattern "Use 'features' to specify a superset of features and use 'exclude' to whittle it down."

**Comment by sparadiso** [comment]
Looking into the issues @bfolie raised.

File: src/citrine/informatics/predictors.py [hunk 5]
```diff
@@ -259,3 +264,119 @@ def _post_dump(self, data: dict) -> dict:
...
+        - XLogP `CDK Link<https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/qsar/descriptors/molecular/XLogPDescriptor.html>`
+        - ChiPath `CDK Link<https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/qsar/descriptors/molecular/ChiPathDescriptor.html>`
+        - FragmentComplexity `CDK Link<https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/qsar/descriptors/molecular/FragmentComplexityDescriptor.html>`
+        - ChiChain `CDK Link<http://cdk.github.io/cdk/2.2/docs/api/org/openscience/cdk/qsar/descriptors/molecular/ChiChainDescriptor.html>`
+        - KierHallSmarts `CDK Link<https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/qsar/descriptors/molecular/KierHallSmartsDescriptor.html>`
+        - MassAutocorr `CDK Link<http://cdk.github.io/cdk/2.2/docs/api/org/openscience/cdk/qsar/descriptors/molecular/AutocorrelationDescriptorMass.html>`
+        - AcidGroupCount `CDK Link<https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/qsar/descriptors/molecular/AcidicGroupCountDescriptor.html>`
+        - BPol `CDK Link<Couldn't find BPol>`
```
Comment (sparadiso) [inline src/citrine/informatics/predictors.py:326]: Missed a link - updating.
